### ISSUE TITLE
Enable source-build pre-built detection

### ISSUE THREAD

**Issue by oleksandr-didyk** [dotnet/roslyn-analyzers#6606]
Part of https://github.com/dotnet/source-build/issues/3017

As part of source-builds team's ongoing work for making .NET source-build sustainable, we are trying to eliminate _pre-builts_ at repo level - dependencies that are not build by source-build or the repo itself (broader definition with examples and description can be found [here](https://github.com/dotnet/source-build/blob/main/Documentation/eliminating-pre-builts.md)). This would be facilitated by existing Arcade build logic that would detect pre-builts as part of the repo's build.

However, before we can enable the logic we first need to resolve existing pre-builts. Full list can be found below:

<details>
<summary>Full list of pre-builts</summary>

```
    <Usage Id="Microsoft.CodeAnalysis" Version="3.8.0" IsDirectDependency="true" />
      <Usage Id="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="3.8.0" />
	  <Usage Id="Microsoft.CodeAnalysis.CSharp" Version="3.8.0" />
	  <Usage Id="Microsoft.CodeAnalysis.Common" Version="3.8.0" />
	    <Usage Id="Microsoft.CodeAnalysis.Analyzers" Version="3.0.0" />
	    <Usage Id="System.Runtime.CompilerServices.Unsafe" Version="4.7.1" />
	  <Usage Id="Microsoft.CodeAnalysis.Workspaces.Common" Version="3.8.0" />
      <Usage Id="Microsoft.CodeAnalysis.VisualBasic.Workspaces" Version="3.8.0" />
	  <Usage Id="Microsoft.CodeAnalysis.VisualBasic" Version="3.8.0" />
	  <Usage Id="Microsoft.CodeAnalysis.Common" Version="3.8.0" />
	    <Usage Id="Microsoft.CodeAnalysis.Analyzers" Version="3.0.0" />
	    <Usage Id="System.Runtime.CompilerServices.Unsafe" Version="4.7.1" />
	  <Usage Id="Microsoft.CodeAnalysis.Workspaces.Common" Version="3.8.0" />
	    <Usage Id="Microsoft.Bcl.AsyncInterfaces" Version="1.1.1" />
	    <Usage Id="System.Composition" Version="1.0.31" />

    <Usage Id="Microsoft.CodeAnalysis" Version="3.7.0" IsDirectDependency="true" />
      <Usage Id="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="3.7.0" />
	  <Usage Id="Microsoft.CodeAnalysis.CSharp" Version="3.7.0" />
	  <Usage Id="Microsoft.CodeAnalysis.Common" Version="3.7.0" />
	    <Usage Id="Microsoft.CodeAnalysis.Analyzers" Version="3.0.0" />
	  <Usage Id="Microsoft.CodeAnalysis.Workspaces.Common" Version="3.7.0" />
      <Usage Id="Microsoft.CodeAnalysis.VisualBasic.Workspaces" Version="3.7.0" />
	  <Usage Id="Microsoft.CodeAnalysis.VisualBasic" Version="3.7.0" />
	  <Usage Id="Microsoft.CodeAnalysis.Common" Version="3.7.0" />
	    <Usage Id="Microsoft.CodeAnalysis.Analyzers" Version="3.0.0" />
	  <Usage Id="Microsoft.CodeAnalysis.Workspaces.Common" Version="3.7.0" />
	    <Usage Id="System.Composition" Version="1.0.31" />

    <Usage Id="Microsoft.CodeAnalysis" Version="3.3.1" IsDirectDependency="true" />
      <Usage Id="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="3.3.1" />
	  <Usage Id="Microsoft.CodeAnalysis.CSharp" Version="3.3.1" />
	  <Usage Id="Microsoft.CodeAnalysis.Common" Version="3.3.1" />
	    <Usage Id="Microsoft.CodeAnalysis.Analyzers" Version="2.9.4" />
	    <Usage Id="System.Threading.Tasks.Extensions" Version="4.5.3" />
	  <Usage Id="Microsoft.CodeAnalysis.Workspaces.Common" Version="3.3.1" />
      <Usage Id="Microsoft.CodeAnalysis.VisualBasic.Workspaces" Version="3.3.1" />
	  <Usage Id="Microsoft.CodeAnalysis.VisualBasic" Version="3.3.1" />
	  <Usage Id="Microsoft.CodeAnalysis.Common" Version="3.3.1" />
	    <Usage Id="Microsoft.CodeAnalysis.Analyzers" Version="2.9.4" />
	    <Usage Id="System.Threading.Tasks.Extensions" Version="4.5.3" />
	  <Usage Id="Microsoft.CodeAnalysis.Workspaces.Common" Version="3.3.1" />
	    <Usage Id="System.Composition" Version="1.0.31" />

    <Usage Id="Microsoft.CodeAnalysis.CSharp" Version="3.9.0" IsDirectDependency="true" />
    <Usage Id="Microsoft.CodeAnalysis.VisualBasic" Version="3.9.0" IsDirectDependency="true" />
    <Usage Id="Microsoft.CodeAnalysis.Common" Version="3.9.0" />

    <Usage Id="Microsoft.CodeAnalysis.BannedApiAnalyzers" Version="3.3.4" IsDirectDependency="true" />

    <Usage Id="Microsoft.Net.Compilers.Toolset" Version="4.5.0" IsDirectDependency="true" IsAutoReferenced="true" />
```

</details>

Specific direct dependencies that are marked as pre-builts:

  - [`Microsoft.Net.Compilers.Toolset: 4.5.0`](https://github.com/dotnet/roslyn-analyzers/issues/6618)
  - [`Microsoft.CodeAnalysis: 3.7.0; 3.8.0; 3.9.0`](https://github.com/dotnet/roslyn-analyzers/issues/6619)

CC: @MichaelSimons

**Comment by MichaelSimons** [comment]
cc @mmitche

**Comment by oleksandr-didyk** [comment]
To outline the current issues: 
  - we cannot resolve the `Microsoft.CodeAnalysis` pre-builts ourselves, as in the context of source-build we cannot support use of historic version (normally we would create a reference package, but as mentioned in the description the assembly would be loaded during the build and fail) nor can we outright bump the version to the latest as it causes exceptions
 
Help on this specific topic would be much appreciated from the maintainers

**Comment by oleksandr-didyk** [comment]
Another issue worth mentioning is `System.Composition: 1.0.31`, a transitive dependency of `Microsoft.CodeAnalysis`

The package is loaded during the build and an SBRP for it exists. When a dependency on SBRP is added, the ref assembly for the package is picked up, crashing the build. I circumvented this by forcing the latest version of the package, which might not be desirable

**Comment by sharwell** [comment]
@oleksandr-didyk It's not clear if/how this is a bug with this repository. It also sounds like this issue is an aggregate issue covering several smaller issues. Can we break them down one at a time to determine if/how each would need to be addressed?

**Comment by oleksandr-didyk** [comment]
> @oleksandr-didyk Oleksandr Didyk FTE It's not clear if/how this is a bug with this repository. It also sounds like this issue is an aggregate issue covering several smaller issues. Can we break them down one at a time to determine if/how each would need to be addressed?

Sure thing, I can create separate issues for each + have this one as tracking

**Comment by oleksandr-didyk** [comment]
> > @oleksandr-didyk Oleksandr Didyk FTE It's not clear if/how this is a bug with this repository. It also sounds like this issue is an aggregate issue covering several smaller issues. Can we break them down one at a time to determine if/how each would need to be addressed?
> 
> Sure thing, I can create separate issues for each + have this one as tracking

Created the issues, removed comments with duplicating information + added more context to the issue's description. I apologize for it not being clear in the beginning
